File: M8/Exemplos/App_eXperience_Ionic/capacitor-elgin-m8/android/src/main/java/com/elgin/m8/Paygo.java
package com.elgin.m8;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Handler;
import android.util.Base64;
import android.util.Log;
import android.widget.Toast;

import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

import org.json.JSONException;
import org.json.JSONObject;

import br.com.setis.interfaceautomacao.Cartoes;
import br.com.setis.interfaceautomacao.Confirmacoes;
import br.com.setis.interfaceautomacao.DadosAutomacao;
import br.com.setis.interfaceautomacao.Financiamentos;
import br.com.setis.interfaceautomacao.ModalidadesPagamento;
import br.com.setis.interfaceautomacao.Operacoes;
import br.com.setis.interfaceautomacao.EntradaTransacao;
import br.com.setis.interfaceautomacao.Personalizacao;
import br.com.setis.interfaceautomacao.SaidaTransacao;
import br.com.setis.interfaceautomacao.StatusTransacao;
import br.com.setis.interfaceautomacao.Transacoes;
import br.com.setis.interfaceautomacao.Versoes;
import br.com.setis.interfaceautomacao.ViasImpressao;

public class Paygo {
    private Activity mainActivity;

    private SaidaTransacao mSaidaTransacao;
    private DadosAutomacao mDadosAutomacao;
    private Transacoes mTransacoes;
    private Confirmacoes mConfirmacao;
    private EntradaTransacao mEntradaTransacao = null;
    private Personalizacao mPersonalizacao;
    private Versoes versoes;
    private Handler mHandler;

    private static PluginCall paygoCall;

    public Paygo(Activity activity){
        mainActivity = activity;
        IniciaClassesInterface(false, false, false, false);
        mHandler = new Handler();
    }

    public void setPluginCall(PluginCall call){
        paygoCall = call;
    }

    private final Runnable mostraJanelaResultado = new Runnable() {//region
        public void run() {
            Map<String, Object> mapValues = new HashMap<>();

            String retorno = mSaidaTransacao.obtemMensagemResultado();
            String via_cliente = "";


            if(mSaidaTransacao.obtemInformacaoConfirmacao()) {
                mConfirmacao.informaStatusTransacao(StatusTransacao.CONFIRMADO_AUTOMATICO);
                mTransacoes.confirmaTransacao(mConfirmacao);

                ViasImpressao vias = mSaidaTransacao.obtemViasImprimir();
                System.out.println("VIAS: " + vias.equals("VIA_NENHUMA"));

                //Imprime a via do cliente
                if ( vias == ViasImpressao.VIA_CLIENTE || vias == ViasImpressao.VIA_CLIENTE_E_ESTABELECIMENTO) {
                    via_cliente = mSaidaTransacao.obtemComprovanteGraficoPortador();
                }

                mapValues.put("retorno", retorno);
                mapValues.put("via_cliente", via_cliente);

                optionsReturnPaygo(mapValues);
            }
        }
    };

    public static void optionsReturnPaygo(Map map){
        Map<String, Object> mapValues = new HashMap<>();

        if(map.get("retorno").equals("Transacao autorizada")){
            String imageViaBase64 = (String) map.get("via_cliente");

           // byte[] decodedString = Base64.decode(imageViaBase64, Base64.DEFAULT);
          //  Bitmap decodedByte = BitmapFactory.decodeByteArray(decodedString, 0, decodedString.length);
           // imageViewViaPaygo.setImageBitmap(decodedByte);

            mapValues.put("quant", 10);
            mapValues.put("base64", imageViaBase64);

            JSObject ret = new JSObject();
            ret.put("via_cliente", imageViaBase64.replaceAll("\\s",""));

            m8PluginPlugin.printerInstance.imprimeCupomTEF(mapValues);
            m8PluginPlugin.printerInstance.AvancaLinhas(mapValues);
            m8PluginPlugin.printerInstance.cutPaper(mapValues);


            paygoCall.resolve(ret);
        }else{
            m8PluginPlugin.alertMessageStatus("Alert", map.get("retorno").toString());
            paygoCall.reject("PayGoError: " + map.get("retorno").toString());
        }

    }



    public void mostraResultadoTransacao(){
        String retorno = "";
        retorno = mSaidaTransacao.obtemMensagemResultado();

        if(mSaidaTransacao.obtemInformacaoConfirmacao()) {
            mConfirmacao.informaStatusTransacao(StatusTransacao.CONFIRMADO_AUTOMATICO);
            mTransacoes.confirmaTransacao(mConfirmacao);

            ViasImpressao vias = mSaidaTransacao.obtemViasImprimir();
            System.out.println("VIAS: " + vias.equals("VIA_NENHUMA"));

            //Imprime a via do cliente
            if ( vias == ViasImpressao.VIA_CLIENTE || vias == ViasImpressao.VIA_CLIENTE_E_ESTABELECIMENTO) {
                String via_cliente = mSaidaTransacao.obtemComprovanteGraficoPortador();
                retorno = via_cliente;
            }
        }
    }


    public void efetuaTransacao(Operacoes operacao, Map map){
        mEntradaTransacao = new EntradaTransacao(operacao, "1");

        if(!operacao.equals(Operacoes.ADMINISTRATIVA)){
            String valor = (String) map.get("valor");
            int parcelas = (Integer) map.get("parcelas");
            String formaPagamento = (String) map.get("formaPagamento");
            String tipoParcelamento = (String) map.get("tipoParcelamento");

            mEntradaTransacao.informaValorTotal(valor);

            if (operacao.equals(Operacoes.VENDA)) {
                mEntradaTransacao.informaDocumentoFiscal("1000");
            }

            mEntradaTransacao.informaModalidadePagamento(ModalidadesPagamento.PAGAMENTO_CARTAO);

            if(formaPagamento.equals("Crédito")){
                mEntradaTransacao.informaTipoCartao(Cartoes.CARTAO_CREDITO);
                mEntradaTransacao.informaNumeroParcelas(parcelas);

            }else if(formaPagamento.equals("Débito")){
                mEntradaTransacao.informaTipoCartao(Cartoes.CARTAO_DEBITO);
                mEntradaTransacao.informaNumeroParcelas(parcelas);

            }else{
                mEntradaTransacao.informaTipoCartao(Cartoes.CARTAO_DESCONHECIDO);
            }

            if(tipoParcelamento.equals("Loja")){
                mEntradaTransacao.informaTipoFinanciamento(Financiamentos.PARCELADO_ESTABELECIMENTO);

            }else if(tipoParcelamento.equals("Adm")){
                mEntradaTransacao.informaTipoFinanciamento(Financiamentos.PARCELADO_EMISSOR);

            }else{
                mEntradaTransacao.informaTipoFinanciamento(Financiamentos.A_VISTA);
            }

            mEntradaTransacao.informaNomeProvedor("DEMO");
            mEntradaTransacao.informaCodigoMoeda("986");
        }

        mConfirmacao = new Confirmacoes();

        new Thread(() -> {
            try {
                mDadosAutomacao.obtemPersonalizacaoCliente();
                mSaidaTransacao = mTransacoes.realizaTransacao(mEntradaTransacao);

                if(mSaidaTransacao == null)
                    return;

                mConfirmacao.informaIdentificadorConfirmacaoTransacao(
                        mSaidaTransacao.obtemIdentificadorConfirmacaoTransacao()
                );

                mEntradaTransacao = null;
            } catch (Exception e) {
                System.out.println("Exception" + e);
            }finally {
                mEntradaTransacao = null;
                if(mSaidaTransacao != null) {
                    mHandler.post(mostraJanelaResultado);
                }
            }
        }).start();
    }

    private void IniciaClassesInterface( boolean suportaViasDiferenciadas, boolean suportaViasReduzidas, boolean troco, boolean desconto ) {
        String versaoAutomacao;
        try {
            versaoAutomacao = mainActivity.getPackageManager().getPackageInfo(
                    mainActivity.getPackageName(), 0).versionName;
        } catch (PackageManager.NameNotFoundException e) {
            versaoAutomacao = "Indisponivel";
        }

        mPersonalizacao = setPersonalizacao(false);

        mDadosAutomacao = new DadosAutomacao("Automacao Demo", versaoAutomacao,"SETIS",
                troco, desconto, suportaViasDiferenciadas, suportaViasReduzidas, true, mPersonalizacao);

        mTransacoes = Transacoes.obtemInstancia(mDadosAutomacao, mainActivity);
        versoes = mTransacoes.obtemVersoes();
    }

    private Personalizacao setPersonalizacao(boolean isInverse) {
        Personalizacao.Builder pb = new Personalizacao.Builder();

        try {
            if (isInverse) {
                pb.informaCorFonte( "#000000" );
                pb.informaCorFonteTeclado("#000000");
                pb.informaCorFundoCaixaEdicao("#FFFFFF");
                pb.informaCorFundoTela("#F4F4F4");
                pb.informaCorFundoTeclado("#F4F4F4");
                pb.informaCorFundoToolbar("#2F67F4");
                pb.informaCorTextoCaixaEdicao("#000000");
                pb.informaCorTeclaPressionadaTeclado("#e1e1e1");
                pb.informaCorTeclaLiberadaTeclado("#dedede");
                pb.informaCorSeparadorMenu("#2F67F4");
            }
        } catch (IllegalArgumentException e) {
            Toast.makeText(mainActivity, "Verifique valores de\nconfiguração", Toast.LENGTH_SHORT).show();
        }
        return pb.build();
    }
}

File: Exemplos/App_eXperience_XamarinAndroid/Tef/PayGo.cs
using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using BR.Com.Setis.Interfaceautomacao;
using Android.Content.PM;

using System.Threading;

namespace M8
{
    class PayGo
    {
        private static Activity mainActivity;

        private static SaidaTransacao mSaidaTransacao;
        private static DadosAutomacao mDadosAutomacao;
        private static Transacoes mTransacoes;
        private static Confirmacoes mConfirmacao;
        private static EntradaTransacao mEntradaTransacao = null;
        private static Personalizacao mPersonalizacao;
        private static Versoes versoes;
        private static Handler mHandler;

       

        public PayGo()
        {
        }

        public static void setActivity(Activity activity)
        {
            mainActivity = activity;
            IniciaClassesInterface(false, false, false, false);
#pragma warning disable CS0618 // O tipo ou membro é obsoleto
            mHandler = new Handler();
#pragma warning restore CS0618 // O tipo ou membro é obsoleto
        }
        private static readonly Java.Lang.Runnable mostraJanelaResultado = new Java.Lang.Runnable(new Action(delegate
        {
            Dictionary<String, String> mapValues = new Dictionary<string, string>();

            String retorno = mSaidaTransacao.ObtemMensagemResultado();
            String via_cliente = "";


            if (mSaidaTransacao.ObtemInformacaoConfirmacao())
            {
                mConfirmacao.InformaStatusTransacao(StatusTransacao.ConfirmadoAutomatico);
                mTransacoes.ConfirmaTransacao(mConfirmacao);



                ViasImpressao vias = mSaidaTransacao.ObtemViasImprimir();
                Console.WriteLine("VIAS: " + vias.Equals("VIA_NENHUMA"));

                //Imprime a via do cliente
                if (vias == ViasImpressao.ViaCliente || vias == ViasImpressao.ViaClienteEEstabelecimento)
                {
                    via_cliente = mSaidaTransacao.ObtemComprovanteGraficoPortador();
                }

                mapValues.Add("retorno", retorno);
                mapValues.Add("via_cliente", via_cliente);



                Tef.OptionsReturnPaygo(mapValues);
            }
        }));

        public void mostraResultadoTransacao()
        {
            String retorno = "";
            retorno = mSaidaTransacao.ObtemMensagemResultado();

            if (mSaidaTransacao.ObtemInformacaoConfirmacao())
            {
                mConfirmacao.InformaStatusTransacao(StatusTransacao.ConfirmadoAutomatico);
                mTransacoes.ConfirmaTransacao(mConfirmacao);

                ViasImpressao vias = mSaidaTransacao.ObtemViasImprimir();
                Console.WriteLine("VIAS: " + vias.Equals("VIA_NENHUMA"));

                //Imprime a via do cliente
                if (vias == ViasImpressao.ViaCliente || vias == ViasImpressao.ViaClienteEEstabelecimento)
                {
                    String via_cliente = mSaidaTransacao.ObtemComprovanteGraficoPortador();
                    retorno = via_cliente;
                }
            }
        }





        public static void efetuaTransacao(Operacoes operacao, Dictionary<String, String> dictionary)
        {
            mEntradaTransacao = new EntradaTransacao(operacao, "1");
            
            if (!operacao.Equals(Operacoes.Administrativa))
            {
                String valor = (String)dictionary["valor"];
                int parcelas = Int32.Parse(dictionary["parcelas"]);
                String formaPagamento = (String)dictionary["formaPagamento"];
                String tipoParcelamento = (String)dictionary["tipoParcelamento"];

                mEntradaTransacao.InformaValorTotal(valor);

                if (operacao.Equals(Operacoes.Venda))
                {
                    mEntradaTransacao.InformaDocumentoFiscal("1000");
                }

                mEntradaTransacao.InformaModalidadePagamento(ModalidadesPagamento.PagamentoCartao);

                if (formaPagamento.Equals("Crédito"))
                {
                    mEntradaTransacao.InformaTipoCartao(Cartoes.CartaoCredito);
                    mEntradaTransacao.InformaNumeroParcelas(parcelas);

                }
                else if (formaPagamento.Equals("Débito"))
                {
                    mEntradaTransacao.InformaTipoCartao(Cartoes.CartaoDebito);
                    mEntradaTransacao.InformaNumeroParcelas(parcelas);

                }
                else
                {
                    mEntradaTransacao.InformaTipoCartao(Cartoes.CartaoDesconhecido);
                }

                if (tipoParcelamento.Equals("Loja"))
                {
                    mEntradaTransacao.InformaTipoFinanciamento(Financiamentos.ParceladoEstabelecimento);

                }
                else if (tipoParcelamento.Equals("Adm"))
                {
                    mEntradaTransacao.InformaTipoFinanciamento(Financiamentos.ParceladoEmissor);
                }
                else
                {
                    mEntradaTransacao.InformaTipoFinanciamento(Financiamentos.AVista);
                }

                mEntradaTransacao.InformaNomeProvedor("DEMO");
                mEntradaTransacao.InformaCodigoMoeda("986");
            }

            mConfirmacao = new Confirmacoes();
            
            new Thread( ()=> {
            try
            {
                mDadosAutomacao.ObtemPersonalizacaoCliente();
                mSaidaTransacao = mTransacoes.RealizaTransacao(mEntradaTransacao);

                if (mSaidaTransacao == null)
                    return;

                mConfirmacao.InformaIdentificadorConfirmacaoTransacao(
                        mSaidaTransacao.ObtemIdentificadorConfirmacaoTransacao()
                );

                mEntradaTransacao = null;
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception" + e);
            }
            finally
            {
                mEntradaTransacao = null;
                if (mSaidaTransacao != null)
                {
                    mHandler.Post(mostraJanelaResultado);
                }
          
          }
        }).Start();
    }





    private static void IniciaClassesInterface(bool suportaViasDiferenciadas, bool suportaViasReduzidas, bool troco, bool desconto)
        {
            System.String versaoAutomacao;
            try
            {
                //versaoAutomacao = mainActivity.getPackageManager().getPackageInfo(
                       // mainActivity.getPackageName(), 0).versionName;
                versaoAutomacao = mainActivity.PackageManager.GetPackageInfo(mainActivity.PackageName, 0).VersionName;
            }
            catch (PackageManager.NameNotFoundException e)
            {
                Console.WriteLine("Exception" + e);
                versaoAutomacao = "Indisponivel";
            }

            mPersonalizacao = setPersonalizacao(false);

            mDadosAutomacao = new DadosAutomacao("Automacao Demo", versaoAutomacao, "SETIS",
                    troco, desconto, suportaViasDiferenciadas, suportaViasReduzidas, true, mPersonalizacao);

            mTransacoes = Transacoes.ObtemInstancia(mDadosAutomacao, mainActivity);
            versoes = mTransacoes.ObtemVersoes();
        }
        private static Personalizacao setPersonalizacao(bool isInverse)
        {
            Personalizacao.Builder pb = new Personalizacao.Builder();
            
            try
            {
                if (isInverse)
                {
                    pb.InformaCorFonte("#000000");
                    pb.InformaCorFonteTeclado("#000000");
                    pb.InformaCorFundoCaixaEdicao("#FFFFFF");
                    pb.InformaCorFundoTela("#F4F4F4");
                    pb.InformaCorFundoTeclado("#F4F4F4");
                    pb.InformaCorFundoToolbar("#2F67F4");
                    pb.InformaCorTextoCaixaEdicao("#000000");
                    pb.InformaCorTeclaPressionadaTeclado("#e1e1e1");
                    pb.InformaCorTeclaLiberadaTeclado("#dedede");
                    pb.InformaCorSeparadorMenu("#2F67F4");
                }
            }
            catch (Java.Lang.IllegalArgumentException)
            {
                Toast.MakeText(mainActivity, "Verifique valores de\nconfiguração", ToastLength.Short).Show();
            }
            return pb.Build();
        }
    }
}